keps/sig-scheduling/3902-decoupled-taint-manager/README.md

```diff
@@ -0,0 +1,305 @@
+    * No
+* **Will enabling / using this feature result in increasing time taken by any operations covered by existing SLIs/SLOs?**
+    * No
+* **Will enabling / using this feature result in non-negligible increase of resource usage (CPU, RAM, disk, IO, ...) in any components?**
+    * No
+
+### Troubleshooting
+
+* **How does this feature react if the API server and/or etcd is unavailable?**
+    * The same behavior as the current version: fail to add taint and/or evict pods on tainted nodes 
+* **What are other known failure modes?**
+    * No
+* **What steps should be taken if SLOs are not being met to determine the problem?**
+    * If the default or custom taint-manager is working or not. 
+
+## Implementation History
+
+* 2023-03-06: Initial KEP published. 
+
+## Drawbacks
+
+## Alternatives
+
+Overall, we need a flexible extension mechanism in eviction through which we can make a call to external systems and interact with other controllers and workload management systems, which have more knowledge about workload disruption. The extension will enable to handle workloads that are split across Kubernetes and Non-Kubernetes platforms.
+
+* **Leverage the existing tolerations for `NoExecute` taints**: operators opt-out from taint-based eviction by injecting tolerations for NoExecute taints. It has a few major disadvantages over an independent `TaintManager`.
+    * It will require to either inject tolerations to all pods via mutating webhooks and/or changing the manifests of all the running workloads. 
+    * It will require either injecting tolerations to every single pod via mutating webhooks and/or changing the manifests of all the running workloads.
+    * It will still need an additional controller to really evict after coordinating with the external reservation/lease system. This will add additional overhead to toleration injections.
+    * Custom mutating webhook may interfere with the admission plugin that applies the default tolerations if they are trying to apply tolerations for similar taints.
+    * Toleration update in Webhooks can also cause conflicts with older version of pods especially during upgrades and we need to ensure a toleration update is valid.
+
+* **Enhance the existing toleration APIs**
+    * Introduce a new Toleration effect `EvictIfAllowed` similar to `PreferNoSchedule` with additional fields on each Toleration to specify a reference to another CR/API object for how to reach the external system that governs if eviction is allowed similar to webhooks. Similar to the communication between API Server and admission webooks, `NoExecTaintManager` and an external service can establish whether eviction for a given pod is allowed or not.
+    * Add RBAC for tolerations API on who can add what tolerations. Otherwise, a user can add tolerations to their workloads(pods) and make every pod land on the node to cause a noisy neighbor and even DoS problem in the worst case.
```

Comment:
I think this is a separated problem than the one we are aiming to solve with this proposal. While there is definitely room for extending the Kubernetes API to support a more graceful form of `NoExecute` respecting the PodDisruptionBudget API (we had a session on this at the Contributor Summit in Amsterdam last week), the goal of this proposal is to ensure operators can extend Kubernetes by providing custom logic handling Taint Based Eviction to change the default behaviour vs having to patch Kubernetes Controller Manager.

Reply:
> I think this is a separated problem than the one we are aiming to solve with this proposal. While there is definitely room for extending the Kubernetes API to support a more graceful form of `NoExecute` respecting the PodDisruptionBudget API (we had a session on this at the Contributor Summit in Amsterdam last week), the goal of this proposal is to ensure operators can extend Kubernetes by providing custom logic handling Taint Based Eviction to change the default behaviour vs having to patch Kubernetes Controller Manager.

Added the comment to the discussion of alternatives.

Reply:
The use and enhancement of tolerations by injecting tolerations for NoExecute is a not viable at scale because of the factors described above.

Although extending the Kubernetes API to support a more graceful form of NoExecute respecting the PodDisruptionBudget API is possible, this proposal aims to enable operators to extend Kubernetes by providing custom logic for handling taint-based evictions. This approach changes the default behavior rather than requiring patching the Kubernetes Controller Manager.

Reply:
Splitting the components is not only for cleanup, but also a prerequisite for enhancing the pod eviction manager, whether it's the native or a custom controller.

Regarding your point about enhancing the default controller, we believe it would be helpful, but we also recognize that it may not be practical or even possible to have a general controller that meets all needs in different use cases, as described in the motivation and use cases.

We will make further clarification in the proposal. 

Thank you for taking the time to review our KEP and share your thoughts.

Reply:
this is what was discussed at SIG Node meeting as well. It will be great to capture improvements planned or desired and pain points and explain how decoupling will help to address those.

Reply:
We summarize the benefits and improvements by decoupling `TaintManager` from `NodeLifecycleController`.

- A more logical code structure with two separate controllers for tainting nodes and evicting pods respectively. 
- Greater flexibility to support customized taint-based eviction to meet the specific needs of different workloads than the use of the toleration-based mechanism, which is less flexible and harder to customize.
- Ability to enable/disable the default `TaintManager` by setting or unsetting the "--controllers" flag of `kube-controller-manager`.

Reply:
I think what @yuanchen8911  is suggesting is decoupling will make it easy for the future changes which are listed above. For example, once the split happens, it would be easier to make a HTTP call to external system which has more context and knowledge of the workloads from the taint controller(from an implementation standpoint).

Reply:
> > the goal of this proposal is to ensure operators can extend Kubernetes by providing custom logic handling Taint Based Eviction to change the default behaviour.
> 
> You can already do that by disabling the controller.
> 
> I don't see any problem with splitting the existing controller into two. This is a rather welcomed cleanup.
> 
> I disagree that the motivation for splitting it is so that people can write their own. We should be looking for opportunities to enhance the existing controller and API so that people don't feel the need to fork it.

I agree splitting will enable to enhance and improve the existing controllers, but it will also provide opportunities for cluster admins to write their own taint managers like scheduler plugins.